File: status_text.py
"""用量文案（无 GUI 依赖，菜单栏进程可安全导入）。"""

from __future__ import annotations

from datetime import datetime, timezone

from cursor_api import (
    UsageSnapshot,
    format_membership_type,
    format_spend_range,
    format_token_count,
)


def format_summary_text(
    usage: UsageSnapshot | None,
    error_message: str | None,
    updated_at: str | None,
    account_label: str | None = None,
) -> str:
    if error_message:
        return f"状态: {error_message} | 更新 {updated_at or '—'}"
    if usage is None:
        return "状态: 等待刷新…"
    auto = "—" if usage.auto_percent_used is None else f"{usage.auto_percent_used:.1f}%"
    api = "—" if usage.api_percent_used is None else f"{usage.api_percent_used:.1f}%"
    est = format_estimated_days(usage)
    tokens = ""
    if usage.total_tokens:
        tokens = f"消耗 {format_token_count(usage.total_tokens)} Token | "
    spend = ""
    if usage.shows_amount():
        spend = f"金额 {format_spend_range(usage.used_cents, usage.limit_cents)} | "
    plan = format_plan_caption(usage.membership_type, account_label)
    if usage.is_unlimited:
        plan = f"{plan} · 不限量"
    return (
        f"剩余 {usage.remaining_percent:.1f}% | {plan} | "
        f"{spend}{tokens}First-party {auto} | API {api} | 预计可用 {est} | 更新 {updated_at or '—'}"
    )


def format_estimated_days(usage: UsageSnapshot) -> str:
    est = usage.estimated_usable_days
    if est is None:
        if usage.used_percent < 0.2:
            return "用量过低，暂无法估算"
        if usage.days_elapsed is not None and usage.days_elapsed < 0.04:
            return "周期刚开始，统计中"
        return "暂无法估算"

    if est <= 0:
        text = "已耗尽"
    elif est < 1:
        text = f"约 {max(1, int(est * 24))} 小时"
    else:
        text = f"约 {est:.1f} 天".replace(".0 天", " 天")

    reset_left = usage.days_remaining
    if reset_left is not None and est > 0:
        if est >= reset_left:
            text += "  ·  可撑过本周期"
        else:
            text += "  ·  可能提前耗尽"
    return text


def status_pill_text(remaining: float | None, *, error: bool = False) -> str:
    """组合 4 左侧状态胶囊。"""
    if error:
        return "异常"
    if remaining is None:
        return "等待刷新"
    pct = float(remaining)
    if pct <= 0:
        return "已耗尽"
    if pct < 20:
        return "额度紧张"
    if pct < 50:
        return "略偏低"
    return "状态良好"


def format_plan_caption(membership: str | None, account_label: str | None = None) -> str:
    raw = (membership or "").strip()
    if not raw:
        name = "—"
    else:
        name = format_membership_type(raw)
        if "套餐" not in name:
            name = f"{name} 套餐"
    label = (account_label or "").strip()
    known = {name.lower(), raw.lower(), format_membership_type(raw).lower()}
    if label and label.lower() not in known:
        if name == "—":
            return label
        return f"{label} · {name}"
    return name


def format_estimate_caption(usage: UsageSnapshot) -> str:
    text = format_estimated_days(usage)
    if "可撑过本周期" in text:
        return "预计可撑过本周期"
    if "提前耗尽" in text:
        return "预计可能提前耗尽"
    if text == "已耗尽":
        return "额度已耗尽"
    return text


def format_reset_date(iso_value: str, include_time: bool = False) -> str:
    try:
        text = iso_value.replace("Z", "+00:00")
        dt = datetime.fromisoformat(text)
        if include_time and dt.tzinfo is not None:
            dt = dt.astimezone()
        text = f"{dt.month}月{dt.day}日"
        if include_time and (dt.hour or dt.minute):
            text += f" {dt.hour:02d}:{dt.minute:02d}"
        return text
    except ValueError:
        return iso_value


def format_cycle_remaining(end_iso: str | None, days_remaining: int | None, now: datetime | None = None) -> str:
    if not end_iso:
        return f"还剩 {days_remaining} 天" if days_remaining is not None else ""
    try:
        text = end_iso.replace("Z", "+00:00")
        end = datetime.fromisoformat(text)
        if end.tzinfo is None:
            end = end.replace(tzinfo=timezone.utc)
        clock = now or datetime.now(timezone.utc)
        if clock.tzinfo is None:
            clock = clock.replace(tzinfo=timezone.utc)
        delta = end - clock.astimezone(end.tzinfo)
    except ValueError:
        return f"还剩 {days_remaining} 天" if days_remaining is not None else ""
    seconds = delta.total_seconds()
    if seconds <= 0:
        return "已到期"
    hours = int(seconds // 3600)
    if hours < 24:
        if hours < 1:
            minutes = max(1, int(seconds // 60))
            return f"还剩 {minutes} 分钟"
        return f"还剩 {hours} 小时"
    days = days_remaining if days_remaining is not None else int(seconds // 86400)
    return f"还剩 {days} 天"


def cycle_end_label(usage: UsageSnapshot) -> str:
    return "到期" if usage.billing_cycle_end_overridden else "重置"


def build_status_lines(
    usage: UsageSnapshot | None,
    error_message: str | None,
    updated_at: str | None = None,
    account_label: str | None = None,
) -> list[tuple[str, str]]:
    """状态明细行（Windows 飞出层 / macOS 原生面板共用）。"""
    if error_message:
        return [("状态", error_message)]
    if usage is None:
        return [("状态", "等待刷新…")]

    rows: list[tuple[str, str]] = []
    if usage.is_unlimited:
        rows.append(("剩余", "不限量"))
    elif usage.shows_amount():
        rows.append(
            (
                "剩余",
                f"{usage.remaining_percent:.1f}%（{format_spend_range(usage.used_cents, usage.limit_cents)}）",
            )
        )
    else:
        rows.append(("剩余", f"{usage.remaining_percent:.1f}%（已用 {usage.used_percent:.1f}%）"))
    label = (account_label or "").strip()
    memb = format_membership_type(usage.membership_type) if usage.membership_type else ""
    if label and label.lower() != memb.lower():
        rows.append(("账号", label))
    plan = memb or "—"
    if usage.is_unlimited:
        plan = f"{plan} · 不限量"
    rows.append(("计划", plan))
    if usage.shows_amount():
        rows.append(("金额", format_spend_range(usage.used_cents, usage.limit_cents)))
    if (
        usage.pooled_used_cents is not None
        and usage.pooled_limit_cents is not None
        and usage.pooled_limit_cents > 0
        and (
            usage.used_cents != usage.pooled_used_cents
            or usage.limit_cents != usage.pooled_limit_cents
        )
    ):
        rows.append(("团队额度", format_spend_range(usage.pooled_used_cents, usage.pooled_limit_cents)))
    if (
        usage.on_demand_used_cents is not None
        and usage.on_demand_limit_cents is not None
        and usage.on_demand_limit_cents > 0
        and (
            usage.used_cents != usage.on_demand_used_cents
            or usage.limit_cents != usage.on_demand_limit_cents
        )
    ):
        rows.append(
            ("按需用量", format_spend_range(usage.on_demand_used_cents, usage.on_demand_limit_cents))
        )
    if usage.total_tokens:
        rows.append(("消耗 Token", format_token_count(usage.total_tokens)))
    if usage.auto_percent_used is not None or usage.api_percent_used is not None:
        auto = "—" if usage.auto_percent_used is None else f"{usage.auto_percent_used:.1f}%"
        api = "—" if usage.api_percent_used is None else f"{usage.api_percent_used:.1f}%"
        rows.append(("明细", f"First-party {auto} · API {api}"))

    if usage.billing_cycle_end:
        end_text = format_reset_date(
            usage.billing_cycle_end, include_time=usage.billing_cycle_end_overridden
        )
        remaining = format_cycle_remaining(usage.billing_cycle_end, usage.days_remaining)
        label = cycle_end_label(usage)
        if remaining:
            rows.append((label, f"{end_text}（{remaining}）"))
        else:
            rows.append((label, end_text))
        rows.append(("预计可用", format_estimated_days(usage)))
    elif usage.estimated_usable_days is not None:
        rows.append(("预计可用", format_estimated_days(usage)))

    rows.append(("更新", updated_at or datetime.now().strftime("%H:%M:%S")))
    return rows

File: macos/Sources/CursorTokenCore/StatusText.swift
import Foundation

public enum StatusText {
    public static func formatSummary(
        _ usage: UsageSnapshot?,
        errorMessage: String?,
        updatedAt: String?,
        accountLabel: String? = nil
    ) -> String {
        if let errorMessage {
            return "状态: \(errorMessage) | 更新 \(updatedAt ?? "—")"
        }
        guard let usage else { return "状态: 等待刷新…" }
        let auto = usage.autoPercentUsed.map { String(format: "%.1f%%", $0) } ?? "—"
        let api = usage.apiPercentUsed.map { String(format: "%.1f%%", $0) } ?? "—"
        let est = formatEstimatedDays(usage)
        var tokens = ""
        if let total = usage.totalTokens, total > 0 {
            tokens = "消耗 \(UsageParser.formatTokenCount(Double(total))) Token | "
        }
        var spend = ""
        if usage.showsAmount {
            spend = "金额 \(UsageParser.formatSpendRange(used: usage.usedCents, limit: usage.limitCents)) | "
        }
        var plan = formatPlanCaption(usage.membershipType, accountLabel: accountLabel)
        if usage.isUnlimited { plan += " · 不限量" }
        var grok = ""
        if usage.showsGrokBot, let left = usage.grokBotRemainingPercent {
            grok = String(format: " | Grok Bot 剩余 %.1f%%", left)
        }
        return "剩余 \(String(format: "%.1f", usage.remainingPercent))% | \(plan) | \(spend)\(tokens)First-party \(auto) | API \(api)\(grok) | 预计可用 \(est) | 更新 \(updatedAt ?? "—")"
    }

    public static func formatEstimatedDays(_ usage: UsageSnapshot) -> String {
        guard let est = usage.estimatedUsableDays else {
            if usage.usedPercent < 0.2 { return "用量过低，暂无法估算" }
            if let elapsed = usage.daysElapsed, elapsed < 0.04 { return "周期刚开始，统计中" }
            return "暂无法估算"
        }
        var text: String
        if est <= 0 {
            text = "已耗尽"
        } else if est < 1 {
            text = "约 \(max(1, Int(est * 24))) 小时"
        } else {
            text = String(format: "约 %.1f 天", est).replacingOccurrences(of: ".0 天", with: " 天")
        }
        if let resetLeft = usage.daysRemaining, est > 0 {
            if est >= Double(resetLeft) {
                text += "  ·  可撑过本周期"
            } else {
                text += "  ·  可能提前耗尽"
            }
        }
        return text
    }

    public static func statusPillText(_ remaining: Double?, error: Bool = false) -> String {
        if error { return "异常" }
        guard let remaining else { return "等待刷新" }
        if remaining <= 0 { return "已耗尽" }
        if remaining < 20 { return "额度紧张" }
        if remaining < 50 { return "略偏低" }
        return "状态良好"
    }

    public static func formatPlanCaption(_ membership: String?, accountLabel: String? = nil) -> String {
        let raw = (membership ?? "").trimmingCharacters(in: .whitespaces)
        var name: String
        if raw.isEmpty {
            name = "—"
        } else {
            name = UsageParser.formatMembershipType(raw)
            if !name.contains("套餐") { name = "\(name) 套餐" }
        }
        let label = (accountLabel ?? "").trimmingCharacters(in: .whitespaces)
        let known: Set<String> = [
            name.lowercased(),
            raw.lowercased(),
            UsageParser.formatMembershipType(raw).lowercased(),
        ]
        if !label.isEmpty, !known.contains(label.lowercased()) {
            if name == "—" { return label }
            return "\(label) · \(name)"
        }
        return name
    }

    public static func formatEstimateCaption(_ usage: UsageSnapshot) -> String {
        let text = formatEstimatedDays(usage)
        if text.contains("可撑过本周期") { return "预计可撑过本周期" }
        if text.contains("提前耗尽") { return "预计可能提前耗尽" }
        if text == "已耗尽" { return "额度已耗尽" }
        return text
    }

    public static func formatResetDate(_ isoValue: String, includeTime: Bool = false) -> String {
        let text = isoValue.replacingOccurrences(of: "Z", with: "+00:00")
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        var dt = f.date(from: text)
        if dt == nil {
            let f2 = ISO8601DateFormatter()
            f2.formatOptions = [.withInternetDateTime]
            dt = f2.date(from: text) ?? f2.date(from: isoValue)
        }
        guard let dt else { return isoValue }
        let cal = Calendar.current
        let m = cal.component(.month, from: dt)
        let d = cal.component(.day, from: dt)
        var label = "\(m)月\(d)日"
        if includeTime {
            let hour = cal.component(.hour, from: dt)
            let minute = cal.component(.minute, from: dt)
            if hour != 0 || minute != 0 {
                label += String(format: " %02d:%02d", hour, minute)
            }
        }
        return label
    }

    public static func formatCycleRemaining(_ endIso: String?, daysRemaining: Int?, now: Date = Date()) -> String {
        guard let endIso, let end = AccountSync.parseIso(endIso) else {
            if let daysRemaining { return "还剩 \(daysRemaining) 天" }
            return ""
        }
        let seconds = end.timeIntervalSince(now)
        if seconds <= 0 { return "已到期" }
        let hours = Int(seconds / 3600)
        if hours < 24 {
            if hours < 1 {
                let minutes = max(1, Int(seconds / 60))
                return "还剩 \(minutes) 分钟"
            }
            return "还剩 \(hours) 小时"
        }
        let days = daysRemaining ?? Int(seconds / 86_400)
        return "还剩 \(days) 天"
    }

    public static func cycleEndLabel(_ usage: UsageSnapshot) -> String {
        usage.billingCycleEndOverridden ? "到期" : "重置"
    }

    public static func buildStatusLines(
        _ usage: UsageSnapshot?,
        errorMessage: String?,
        updatedAt: String? = nil,
        accountLabel: String? = nil
    ) -> [(String, String)] {
        if let errorMessage { return [("状态", errorMessage)] }
        guard let usage else { return [("状态", "等待刷新…")] }

        var rows: [(String, String)] = []
        if usage.isUnlimited {
            rows.append(("剩余", "不限量"))
        } else if usage.showsAmount {
            rows.append((
                "剩余",
                String(format: "%.1f%%（%@）", usage.remainingPercent, UsageParser.formatSpendRange(used: usage.usedCents, limit: usage.limitCents))
            ))
        } else {
            rows.append(("剩余", String(format: "%.1f%%（已用 %.1f%%）", usage.remainingPercent, usage.usedPercent)))
        }
        let label = (accountLabel ?? "").trimmingCharacters(in: .whitespaces)
        let memb = usage.membershipType.isEmpty ? "" : UsageParser.formatMembershipType(usage.membershipType)
        if !label.isEmpty, label.lowercased() != memb.lowercased() {
            rows.append(("账号", label))
        }
        var plan = memb.isEmpty ? "—" : memb
        if usage.isUnlimited { plan += " · 不限量" }
        rows.append(("计划", plan))
        if usage.showsAmount {
            rows.append(("金额", UsageParser.formatSpendRange(used: usage.usedCents, limit: usage.limitCents)))
        }
        if let pu = usage.pooledUsedCents, let pl = usage.pooledLimitCents, pl > 0,
           usage.usedCents != pu || usage.limitCents != pl
        {
            rows.append(("团队额度", UsageParser.formatSpendRange(used: pu, limit: pl)))
        }
        if let ou = usage.onDemandUsedCents, let ol = usage.onDemandLimitCents, ol > 0,
           usage.usedCents != ou || usage.limitCents != ol
        {
            rows.append(("按需用量", UsageParser.formatSpendRange(used: ou, limit: ol)))
        }
        if let tokens = usage.totalTokens, tokens > 0 {
            rows.append(("消耗 Token", UsageParser.formatTokenCount(Double(tokens))))
        }
        if usage.autoPercentUsed != nil || usage.apiPercentUsed != nil {
            let auto = usage.autoPercentUsed.map { String(format: "%.1f%%", $0) } ?? "—"
            let api = usage.apiPercentUsed.map { String(format: "%.1f%%", $0) } ?? "—"
            rows.append(("明细", "First-party \(auto) · API \(api)"))
        }
        if usage.showsGrokBot, let grokUsed = usage.grokBotPercentUsed {
            var grok = String(format: "剩余 %.1f%%（本周已用 %.1f%%）", usage.grokBotRemainingPercent ?? 0, grokUsed)
            if let reset = usage.grokBotResetAt {
                let resetText = formatResetDate(reset)
                let remaining = formatCycleRemaining(reset, daysRemaining: usage.grokBotDaysRemaining)
                grok += remaining.isEmpty ? " · \(resetText) 重置" : " · \(resetText)（\(remaining)）"
            }
            rows.append(("Grok Bot", grok))
        }
        if let end = usage.billingCycleEnd {
            let endText = formatResetDate(end, includeTime: usage.billingCycleEndOverridden)
            let remaining = formatCycleRemaining(end, daysRemaining: usage.daysRemaining)
            let label = cycleEndLabel(usage)
            if remaining.isEmpty {
                rows.append((label, endText))
            } else {
                rows.append((label, "\(endText)（\(remaining)）"))
            }
            rows.append(("预计可用", formatEstimatedDays(usage)))
        } else if usage.estimatedUsableDays != nil {
            rows.append(("预计可用", formatEstimatedDays(usage)))
        }
        let stamp = updatedAt ?? {
            let f = DateFormatter()
            f.dateFormat = "HH:mm:ss"
            return f.string(from: Date())
        }()
        rows.append(("更新", stamp))
        return rows
    }

    /// Menu-bar template: unconfigured is idle (not error); other failures show "!".
    public static func trayTemplateState(errorMessage: String?, remainingPercent: Double?) -> (remaining: Double?, error: Bool) {
        if let err = errorMessage, err.hasPrefix("未配置") {
            return (nil, false)
        }
        if errorMessage != nil {
            return (nil, true)
        }
        return (remainingPercent, false)
    }

    public static func trayPercentLabel(_ remaining: Double?, error: Bool) -> String {
        if error { return "!" }
        guard let remaining else { return "–" }
        let pct = min(100, max(0, remaining))
        if pct >= 99.5 { return "100" }
        return String(Int(pct.rounded()))
    }
}
